src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -642,5 +717,204 @@ private static void ThrowMarkedCtorDoesNotTakeAllProvidedArguments()
         {
             throw new InvalidOperationException(SR.Format(SR.MarkedCtorMissingArgumentTypes, nameof(ActivatorUtilitiesConstructorAttribute)));
         }
+
+#if NET8_0_OR_GREATER // Use the faster ConstructorInvoker which also has alloc-free APIs when <= 4 parameters.
+        private static object ReflectionFactoryServiceOnlyFixed(
+            ConstructorInvoker invoker,
+            Type[] parameterTypes,
+            Type declaringType,
+            IServiceProvider serviceProvider)
+        {
+            Debug.Assert(parameterTypes.Length >= 1 && parameterTypes.Length <= FixedArgumentThreshold);
+            Debug.Assert(FixedArgumentThreshold == 4);
+
+            if (serviceProvider is null)
+                ThrowHelperArgumentNullExceptionServiceProvider();
```

Comment:
NIT: Since it is within .NET 8 if-def could we use `ArgumentNullException.ThrowIfNull(serviceProvider)`? Here and 4 similar cases below

Reply:
Yeah there was an earlier discussion around this - it's at least consistent now across all cases using the helper method. In one case, under .NET 8, we couldn't use ThrowIfNull due to a compile warning around unused variable (but we still want to verify for consistency) and of course under .NetStandard\.NetFx we can't use ThrowIfNull at all.